File: internal/web/static/notifications.js
// notifications.js — part of gg's web client.
//
// The notification centre: the browser's half of the TUI's `!` dialog.
//
// The TUI notices repository problems it can fix and offers the fix behind one
// key; the browser noticed none of them. The finding carried here today is the
// narrowed fetch refspec — a clone whose refspec cannot map a branch, so a push
// succeeds while the ↓↑ markers never move and nobody is told why.
//
// Everything this file remembers lives SERVER-SIDE, in gg's machine-local
// prompt store. `gg web` binds a random loopback port, so every run is a new
// origin and localStorage starts empty — a dismissal kept in the browser would
// come back on the next run, which is exactly the bug the centre must not have.
import { $, esc, getJSON, postJSON } from "./core.js";
import { closeLayer, mountOverlay, pushLayer, topLayer } from "./layers.js";
import { registerHelp, registerRows } from "./menus.js";
import { opLine, startOp } from "./ops.js";

// Notice ids are PERMANENT — the server remembers a dismissal by id in
// prompts.toml. This mirrors the constant in internal/web/notifications.go and
// must never be renamed on one side only.
const NOTICE_REFSPEC = "narrow_fetch_refspec";

// This sheet has NO global .hidden rule — every surface hides itself by id (a
// new element with class="hidden" and no rule behind it is plainly visible;
// that bug has shipped here before). A feature that mounts its own markup
// therefore brings its own styles, which also keeps style.css out of parallel
// merges.
const css = document.createElement("style");
css.textContent = `
#notice-chip {
  margin-left: 10px; background: var(--sel); color: #f27a6a;
  border: 1px solid #f27a6a; border-radius: 3px; padding: 1px 10px;
  font: inherit; cursor: pointer;
}
#notice-chip:hover { background: #f27a6a; color: var(--bg); }
#notice-chip.hidden { display: none; }
#notify { position: fixed; inset: 0; background: rgba(0,0,0,.55); display: flex; align-items: center; justify-content: center; z-index: 11; cursor: pointer; }
#notify.hidden { display: none; }
#notify-box { background: var(--bg-alt); border: 1px solid var(--accent); border-radius: 6px; padding: 18px 24px; width: 90vw; max-width: 720px; max-height: 84vh; overflow-y: auto; cursor: default; }
#notify-box h2 { margin: 0 0 12px; font-size: 15px; }
#notify-box .nempty { color: var(--dim); }
#notify-box .nnotice { border-top: 1px solid var(--border); padding: 12px 0 4px; }
#notify-box .nnotice:first-of-type { border-top: none; }
#notify-box .nhead { display: flex; align-items: baseline; gap: 12px; }
#notify-box .ntitle { flex: 1 1 auto; }
#notify-box .ndetail { color: var(--dim); font-size: 12px; line-height: 1.6; margin: 6px 0 8px; }
#notify-box .nitem { display: flex; align-items: baseline; gap: 12px; padding: 3px 0 3px 14px; }
#notify-box .nname { flex: 1 1 auto; color: var(--accent); overflow: hidden; text-overflow: ellipsis; white-space: nowrap; }
#notify-box button { background: var(--bg); color: var(--fg); border: 1px solid var(--border); border-radius: 3px; padding: 1px 10px; font: inherit; cursor: pointer; }
#notify-box button:hover { border-color: var(--accent); }
`;
document.head.append(css);

// The bell: a chip in the top bar beside the other two, red because every
// finding it carries is something quietly going wrong. Built here rather than
// in index.html so this feature owns all of its own markup.
const chip = document.createElement("button");
chip.id = "notice-chip";
chip.className = "hidden";
chip.title = "repository notices — click to see what gg can fix";
$("top").insertBefore(chip, $("pull-btn"));
chip.addEventListener("click", openNotifications);

// notices is the last answer from the server: the findings worth showing, each
// already carrying the actions that still apply to it.
let notices = [];

async function fetchNotifications() {
  try {
    const resp = await getJSON("/api/notifications");
    notices = resp.notices || [];
  } catch {
    notices = []; // a centre that cannot load says nothing, it does not shout
  }
  chip.textContent = "! " + notices.length + (notices.length === 1 ? " notice" : " notices");
  chip.classList.toggle("hidden", notices.length === 0);
  if (!$("notify").classList.contains("hidden")) renderNotifications();
  return notices;
}


function openNotifications() {
  renderNotifications();
  pushLayer("notify", $("notify"), {
    onKey: (e) => {
      if (e.key === "Escape" || e.key === "!") closeLayer("notify");
      e.preventDefault();
      return true; // the centre owns the keyboard until closed
    },
  });
  fetchNotifications(); // opening is also a refresh — the cheapest honest one
}


// The overlay mounts itself: no markup in index.html, so several features can
// add surfaces without ever meeting in that file.
const box = mountOverlay("notify");
box.innerHTML = `<div id="notify-box"><h2>notifications</h2><div id="notify-list"></div></div>`;
box.addEventListener("click", (e) => {
  if (e.target === box) closeLayer("notify"); // click-off, like #help
});


function renderNotifications() {
  const list = $("notify-list");
  if (!notices.length) {
    list.innerHTML = `<div class="nempty">nothing to report — gg found no repository problems it can fix.</div>`;
    return;
  }
  list.innerHTML = notices
    .map((n, ni) => {
      const all = n.actions || [];
      const perItem = all.filter((a) => a.branch);
      // data-a is the index into n.actions, NOT into the filtered list the row
      // is rendered from — the click handler resolves against n.actions, and a
      // filtered index there would fire the wrong action.
      const head =
        `<div class="nhead"><span class="ntitle">${esc(n.title)}</span>` +
        all
          .map((a, ai) => (a.branch ? "" : `<button data-n="${ni}" data-a="${ai}">${esc(a.label)}</button>`))
          .join("") +
        `<button data-n="${ni}" data-dismiss="1">dismiss</button></div>`;
      const detail = (n.detail || []).map((d) => `<div>${esc(d)}</div>`).join("");
      const items = (n.items || [])
        .map((it) => {
          const a = perItem.find((x) => x.branch === it);
          const idx = all.indexOf(a);
          const btn = a ? `<button data-n="${ni}" data-a="${idx}">${esc(a.label)}</button>` : "";
          return `<div class="nitem"><span class="nname">${esc(it)}</span>${btn}</div>`;
        })
        .join("");
      return `<div class="nnotice">${head}<div class="ndetail">${detail}</div>${items}</div>`;
    })
    .join("");
}


$("notify-list").addEventListener("click", (e) => {
  const btn = e.target.closest("button");
  if (!btn) return;
  const n = notices[Number(btn.dataset.n)];
  if (!n) return;
  if (btn.dataset.dismiss) {
    dismissNotice(n.id);
    return;
  }
  // The header buttons are rendered from a FILTERED list, so an action is
  // always resolved against n.actions by its real index — never against the
  // filtered one.
  const a = n.actions[Number(btn.dataset.a)];
  if (!a) return;
  if (a.dismiss) {
    dismissNotice(a.dismiss); // a suppression, not an operation: the centre stays open
    return;
  }
  closeLayer("notify"); // the op's progress line is on the page behind this
  const body = { op: a.op };
  if (a.branch) body.branch = a.branch;
  startOp(body, a.label);
});


// dismissNotice persists "never for this repo" through the server. It is the
// TUI's key (the git common dir), so dismissing the shared refspec advice here
// silences the TUI's identical notice too.
async function dismissNotice(id) {
  try {
    await postJSON("/api/notifications/dismiss", { id });
  } catch (e) {
    opLine("error: " + (e.message || e), true);
    return;
  }
  fetchNotifications();
}


// --- refresh ---------------------------------------------------------------
// The centre must not poll. ops.js re-enables the pull/push buttons in exactly
// one place — the terminal `done` event — so that flip from disabled to enabled
// IS "an operation finished", and observing it costs nothing and fires never
// otherwise. (ops.js is shared by every parallel branch and must not be edited,
// which is why the signal is read off the DOM rather than called directly.)
new MutationObserver((records) => {
  for (const r of records) {
    if (r.attributeName === "disabled" && !$("push-btn").disabled) {
      fetchNotifications();
      return;
    }
  }
}).observe($("push-btn"), { attributes: true, attributeFilter: ["disabled"] });


// A branch the centre reports as unmapped can be repaired from its own menu —
// the finding is about that branch, and the row is where the user is already
// looking. Rows register themselves; sidebar.js is never edited.
registerRows("branch", (b) => {
  if (!b || !b.name) return [];
  const n = notices.find((x) => x.id === NOTICE_REFSPEC);
  if (!n || !(n.items || []).includes(b.name)) return [];
  return [
    { sep: true },
    {
      label: "add fetch mapping + fetch",
      act: () => startOp({ op: "add-fetch-mappings", branch: b.name }, "mapping " + b.name),
    },
  ];
});


registerHelp({ key: "!", html: "notifications — repository problems gg can fix (an unmapped fetch refspec today)" });

// The bell answers to `!`, like the TUI's. Registered here (not in keys.js) so
// the key and the surface stay in one file; layers own the keyboard first, so
// this only fires when nothing is open — the keys.js rule.
document.addEventListener("keydown", (e) => {
  if (e.key !== "!" || topLayer()) return;
  if (e.target.closest && e.target.closest("input,textarea")) return;
  e.preventDefault();
  openNotifications();
});

fetchNotifications();

export { fetchNotifications, openNotifications };

File: internal/web/notifications.go
package web

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"sync"

	"github.com/homeend/gigagit/internal/domain"
	"github.com/homeend/gigagit/internal/engine"
	"github.com/homeend/gigagit/internal/model"
)

// The notification centre — the browser's half of the TUI's `!` dialog.
//
// The TUI notices repository problems it can fix and offers the fix behind one
// key; the browser noticed none of them. The first (and so far only) finding
// carried here is the narrowed fetch refspec: on a single-branch or shallow
// clone a push succeeds while the branch's remote-tracking ref never moves, so
// the ↓↑ markers and ahead/behind silently stay wrong and nobody is told why.
//
// Two rules this file lives under:
//
//   - **Notice ids are permanent.** A dismissal is remembered by id in
//     prompts.toml, so an id that ships is a promise to that file: renaming one
//     silently un-dismisses it for everybody. Choose them like config keys.
//   - **Suppression lives in promptstate, never in browser storage.** `gg web`
//     binds a RANDOM loopback port, so every run is a different origin and
//     localStorage starts empty — a dismissal kept there would come back on the
//     next run.

// noticeNarrowRefspec is the "branches aren't tracked by the fetch refspec"
// finding. PERMANENT id, and deliberately the TUI's own (internal/tui/notify.go):
// it is the same advice about the same repo, so dismissing it in either
// frontend silences both — the commit_graph_recommend precedent in health.go.
const noticeNarrowRefspec = "narrow_fetch_refspec"

// noticePushMapOffer suppresses the fork engine.Push already raises after a
// push whose remote-tracking ref did not move (FetchMappingDecisionID): on a
// single-branch clone that question comes back on EVERY push of every unmapped
// branch, and a modal you cannot turn off is one you learn to dismiss without
// reading. Dismissing this id answers it "skip" from then on.
//
// PERMANENT id, and separate from the notice above on purpose: silencing the
// per-push question must not also hide the finding, which stays in the centre
// with its repair — that is what makes turning the question off safe rather
// than a way to lose the problem.
//
// Scoped per repo (DismissNotice) rather than globally: a narrowed refspec is a
// property of one clone, so an answer here says nothing about another repo.
const noticePushMapOffer = "web_push_mapping_offer"

// dismissableNotices is the allowlist for POST /api/notifications/dismiss. A
// client bug can then never write junk keys into prompts.toml, which is a file
// nothing ever garbage-collects (the handleNoticeDismiss precedent).
var dismissableNotices = map[string]bool{
	noticeNarrowRefspec: true,
	noticePushMapOffer:  true,
}

// opAddFetchMappings is the wire name of the refspec repair. PERMANENT.
const opAddFetchMappings = "add-fetch-mappings"

// noticeAction is one offer attached to a notice. Op is the wire op name the
// client POSTs to /api/op; Branch narrows it to a single branch ("" = every
// affected one, the batch action). Dismiss, when set instead of Op, makes the
// row a suppression rather than an operation — the id it names is dismissed.
type noticeAction struct {
	Op      string `json:"op,omitempty"`
	Label   string `json:"label"`
	Branch  string `json:"branch,omitempty"`
	Dismiss string `json:"dismiss,omitempty"`
}

// noticeWire is one finding as the browser sees it.
type noticeWire struct {
	ID      string         `json:"id"`
	Title   string         `json:"title"`
	Detail  []string       `json:"detail"`
	Items   []string       `json:"items,omitempty"` // the affected things, one row each
	Actions []noticeAction `json:"actions"`
}

type noticesResp struct {
	Notices []noticeWire `json:"notices"`
	// Suppressed reports the dismissable ids the user already turned off —
	// including ones that produce no notice row (the post-push offer), which
	// the client needs before deciding whether to pop it up.
	Suppressed map[string]bool `json:"suppressed"`
}

// NoticeBuilder derives zero or more notices from one health snapshot. It is
// the notification centre's extension point: a feature adds a finding from its
// OWN file (RegisterNotice in an init) rather than editing this one, exactly as
// RegisterOp and registerRows keep parallel features out of each other's merges.
// Builders must not perform network calls — the centre is refreshed after every
// operation and has to stay cheap. dismissed carries this repo's persisted
// dismissals so a builder can leave out an offer the user already turned off
// without re-reading the store.
type NoticeBuilder func(ctx context.Context, s *Server, svc *domain.Service, h model.RepoHealth, dismissed map[string]bool) []noticeWire

var noticeRegistry []NoticeBuilder

// RegisterNotice adds a builder to the notification centre.
func RegisterNotice(b NoticeBuilder) {
	if b == nil {
		panic("web: RegisterNotice with a nil builder")
	}
	noticeRegistry = append(noticeRegistry, b)
}

func init() {
	RegisterNotice(narrowRefspecNotice)
	RegisterAutoAnswer(skipSuppressedFetchMapping)
	RegisterRoutes(func(mux *http.ServeMux, s *Server) {
		mux.HandleFunc("GET /api/notifications", s.handleNotifications)
		mux.HandleFunc("POST /api/notifications/dismiss", writeGuard(s.handleNotificationDismiss))
	})
	RegisterOp(opAddFetchMappings, buildAddFetchMappings)
}

// handleNotifications answers with every finding that is neither dismissed for
// this repo nor empty. Health failures degrade to an empty list rather than an
// error: the TUI's "health never surfaces errors in the UI" posture applies
// here too — a centre that 500s is worse than a centre with nothing in it.
func (s *Server) handleNotifications(w http.ResponseWriter, r *http.Request) {
	svc := s.service()
	resp := noticesResp{Notices: []noticeWire{}, Suppressed: map[string]bool{}}
	h, err := svc.RepoHealth(readCtx(r))
	if err != nil {
		writeJSON(w, resp)
		return
	}
	rememberRepoKey(svc, h.GitCommonDir) // for the auto-answer policies, which cannot ask git
	var dismissed map[string]bool
	if store := s.promptStore(); store != nil && h.GitCommonDir != "" {
		dismissed = store.DismissedNotices(h.GitCommonDir)
	}
	for id := range dismissableNotices {
		resp.Suppressed[id] = dismissed[id]
	}
	for _, build := range noticeRegistry {
		for _, n := range build(readCtx(r), s, svc, h, dismissed) {
			if dismissed[n.ID] {
				continue
			}
			resp.Notices = append(resp.Notices, n)
		}
	}
	writeJSON(w, resp)
}

// handleNotificationDismiss persists a "never for this repo" dismissal into the
// TUI-shared prompts store, keyed by git common dir. The id is allowlisted.
func (s *Server) handleNotificationDismiss(w http.ResponseWriter, r *http.Request) {
	var req struct {
		ID string `json:"id"`
	}
	if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, 8<<10)).Decode(&req); err != nil {
		writeErr(w, http.StatusBadRequest, errors.New("bad request body"))
		return
	}
	if !dismissableNotices[req.ID] {
		writeErr(w, http.StatusBadRequest, errors.New("unknown notice id"))
		return
	}
	svc := s.service()
	key, err := svc.GitCommonDir(r.Context())
	if err != nil || key == "" {
		writeErr(w, http.StatusInternalServerError, errors.New("cannot resolve repo key"))
		return
	}
	store := s.promptStore()
	if store == nil {
		writeErr(w, http.StatusInternalServerError, errors.New("no state dir for dismissals"))
		return
	}
	if err := store.DismissNotice(key, req.ID); err != nil {
		writeErr(w, http.StatusInternalServerError, err)
		return
	}
	rememberRepoKey(svc, key)
	writeJSON(w, map[string]bool{"ok": true})
}

// narrowRefspecNotice mirrors the TUI notice of the same name: branches whose
// upstream is configured but which the clone's fetch refspec cannot resolve.
// Each affected branch is its own row with its own repair, plus one batch
// action for all of them — the TUI only offers the batch, and a per-branch fix
// matters here because the browser lists them.
func narrowRefspecNotice(_ context.Context, _ *Server, _ *domain.Service, h model.RepoHealth, dismissed map[string]bool) []noticeWire {
	if len(h.UnmappedBranches) == 0 {
		return nil
	}
	title := "1 branch isn't tracked by the fetch refspec"
	if len(h.UnmappedBranches) != 1 {
		title = fmt.Sprintf("%d branches aren't tracked by the fetch refspec", len(h.UnmappedBranches))
	}
	actions := []noticeAction{{
		Op:    opAddFetchMappings,
		Label: "add mappings + fetch",
	}}
	for _, b := range h.UnmappedBranches {
		actions = append(actions, noticeAction{Op: opAddFetchMappings, Label: "fix " + b, Branch: b})
	}
	// The escape hatch for the fork engine.Push raises after every push of an
	// unmapped branch. Offered only while it still fires, and never as the way
	// to make the problem disappear — the finding and its repair stay right
	// here afterwards.
	if !dismissed[noticePushMapOffer] {
		actions = append(actions, noticeAction{
			Label:   "stop asking after each push",
			Dismiss: noticePushMapOffer,
		})
	}
	return []noticeWire{{
		ID:    noticeNarrowRefspec,
		Title: title,
		Detail: []string{
			"This clone's fetch refspec doesn't map these branches, so a push never moves their remote-tracking ref — the ↓↑ tip markers and ahead/behind cannot follow them.",
			"gg can add a per-branch mapping and fetch just those branches (no mass download).",
		},
		Items:   h.UnmappedBranches,
		Actions: actions,
	}}
}

// The repo key an auto-answer policy needs, cached so the policy never asks git
// for it.
//
// svc.GitCommonDir takes a READ reservation on the repo gate. A decider runs
// inside an operation that already holds a write one, so asking there blocks
// until the operation finishes — and the operation is blocked waiting for the
// answer. That is not theoretical: it hung a push until the decide timeout, and
// the browser's decide POST raced ahead of the parked fork and 409'd.
//
// So the key is remembered by the handlers that resolve it anyway, none of
// which run under an operation, and the policy reads only the cache. One slot,
// keyed by the service pointer: a re-root swaps the service and the stale entry
// with it. An unknown key declines to claim the decision, so the modal opens
// exactly as it does today — fail-open.
var (
	repoKeyMu  sync.Mutex
	repoKeySvc *domain.Service
	repoKeyVal string
)

func rememberRepoKey(svc *domain.Service, key string) {
	if svc == nil || key == "" {
		return
	}
	repoKeyMu.Lock()
	repoKeySvc, repoKeyVal = svc, key
	repoKeyMu.Unlock()
}

func cachedRepoKey(svc *domain.Service) string {
	repoKeyMu.Lock()
	defer repoKeyMu.Unlock()
	if svc != repoKeySvc {
		return ""
	}
	return repoKeyVal
}

// skipSuppressedFetchMapping answers engine.Push's post-push "add a tracking
// mapping?" fork with "skip" once the user has turned that question off for
// this repo. Without it the fork is unanswerable-forever: it fires on every
// push of every unmapped branch, and the browser has nowhere of its own to
// remember an answer (a random port means a fresh localStorage every run).
//
// The only work here is a small TOML read, once per fork.
func skipSuppressedFetchMapping(s *Server, req engine.DecisionRequest) (string, bool) {
	if req.ID != engine.FetchMappingDecisionID {
		return "", false
	}
	key := cachedRepoKey(s.service())
	if key == "" {
		return "", false
	}
	store := s.promptStore()
	if store == nil {
		return "", false
	}
	if !store.DismissedNotices(key)[noticePushMapOffer] {
		return "", false
	}
	return "skip", true
}

// buildAddFetchMappings is the refspec repair. The branch names are resolved
// SERVER-SIDE against a fresh health read and never taken from the wire beyond
// a membership check: they flow into a config write and into `git fetch` argv,
// so a client-named branch must be one gg itself just reported as affected.
// An omitted branch is the batch action — every affected branch at once.
func buildAddFetchMappings(s *Server, r *http.Request, req opStartRequest) (engine.Operation, func(), int, error) {
	svc := s.service()
	h, err := svc.RepoHealth(r.Context())
	if err != nil {
		return nil, nil, http.StatusInternalServerError, err
	}
	if len(h.UnmappedBranches) == 0 {
		return nil, nil, http.StatusUnprocessableEntity, errors.New("no branches need a fetch mapping")
	}
	branches := h.UnmappedBranches
	if req.Branch != "" {
		found := false
		for _, b := range h.UnmappedBranches {
			if b == req.Branch {
				found = true
				break
			}
		}
		if !found {
			return nil, nil, http.StatusUnprocessableEntity, errors.New("branch does not need a fetch mapping")
		}
		branches = []string{req.Branch}
	}
	// Remote "origin" matches the detection: unmappedFromConfig only ever
	// reports branches configured to track origin, precisely so this repair
	// cannot forge a mapping the fetch could not satisfy.
	return engine.AddFetchMappings{Remote: "origin", Branches: branches}, nil, 0, nil
}
